compiler/src/dmd/expression.d

```diff
@@ -7407,6 +7409,68 @@ bool isArrayConstructionOrAssign(const Identifier id)
         id == Id._d_arraysetassign;
 }
 
+/**
+ * Checks if the given expresion is a call to the runtime hook `id`.
+ *
+ * Params:
+ *    e = the expression to check
+ *    id = the identifier of the runtime hook
+ * Returns:
+ *    `e` cast to `CallExp` if it's the hook, `null` otherwise
+ */
+CallExp isRuntimeHook(Expression e, Identifier id)
+{
+    if (auto ce = e.isCallExp())
+    {
+        if (auto ve = ce.e1.isVarExp())
+        {
+            if (auto fd = ve.var.isFuncDeclaration())
+            {
+                // If `_d_HookTraceImpl` is found, resolve the underlying hook
+                // and replace `e` and `fd` with it.
+                removeHookTraceImpl(ce, fd);
+                return fd.ident == id ? ce : null;
+            }
+        }
+    }
+
+    return null;
+}
+
+/**
+ * Removes `_d_HookTraceImpl` if found from `ce` and `fd`.
+ * This is needed for the CTFE interception code to be able to find hooks that are called though the hook's `*Trace`
+ * wrapper.
+ *
+ * This is done by replacing `_d_HookTraceImpl!(T, Hook, errMsg)(..., parameters)` with `Hook(parameters)`.
+ * Parameters:
+ *  ce = The CallExp that possible will be be replaced
+ *  fd = Fully resolve function declaration that `ce` would call
+ */
+void removeHookTraceImpl(ref CallExp ce, ref FuncDeclaration fd)
```

Comment:
Why are these functions being moved  from dinterpret?

Reply:
Because I'm using `removeHookTraceImpl()` both here and in `dinterpret.d` [1]. And I'm using `isRuntimeHook()` both in `dinterpred.d` [2] and in `expressiosnem.d` [3].

[1] https://github.com/teodutu/dmd/blob/template-_d_arraycatnTX/compiler/src/dmd/dinterpret.d#L4737
[2] https://github.com/teodutu/dmd/blob/29cd6681feb448ce40ed689397dbe4a223f22e82/compiler/src/dmd/dinterpret.d#L999
[3] https://github.com/teodutu/dmd/blob/29cd6681feb448ce40ed689397dbe4a223f22e82/compiler/src/dmd/expressionsem.d#L10966